### ISSUE TITLE
【社区任务】Liger Softmax 性能优化

### ISSUE THREAD

**Issue by stwilliam503388-creator** [triton-lang/triton-ascend#1802]
## 认领方式

在当前Issue下留言「认领这个任务」
我们会及时分配并@您
一位开发者只能同时领取一个任务，启动验收后才能领取下一个，同时需要开发者在昇腾社区任务广场的对应任务完成报名
开发者需要定期和对接人沟通进展，除特殊情况外，如果连续两个星期没有进展，任务将被释放

## 社区任务广场

[【Triton-Ascend】Liger Softmax 性能优化](https://www.hiascend.com/activities/task-center/details/9c128e544de24906b1bc961e89647606)

---

## 任务类型

算子开发

## 任务描述

在昇腾 NPU 上基于 Triton-Ascend 编程语言，完成 [liger-kernel/liger-kernel](https://github.com/linkedin/Liger-Kernel) 中 LigerSoftmax 算子的昇腾亲和支持。Softmax 是基础激活/归一化算子，在 attention 等场景高频使用。完成算子设计、开发、测试及性能优化全流程工作，验收通过后将算子提交至昇腾算子开源仓。

## 任务标准

1. 在 liger_kernel 0.8.2 + triton-ascend 3.2.2 + torch / torch_npu（性能数据采集自 Ascend 910B3 平台，baseline 为 torch 原生实现），运行 benchmark/scripts/benchmark_softmax.py 全部配置（forward / backward / full 各 kernel_operation_mode）正常产出数据，无崩溃、无超时；
2. 性能达标：所有配置下 liger kernel 耗时均不慢于 baseline（ratio ≤ 1.0），修复后提交优化前后逐配置对比数据表；
3. 平台覆盖：在 Ascend A2、A3、950 平台上均可运行且满足上述性能标准；如因平台硬件差异存在限制，须在 PR 中说明原因；
4. 精度不回归：该算子对应的 pytest 单测（含 fwd/bwd 数值与梯度检查）全部通过；
5. 提交 PR，附 profiling 根因分析、优化手段说明（如 tile/block 划分、launch 开销、非融合路径等）与复测命令及结果。

### 功能实现要求

1. 基于 Triton-Ascend 3.2.2 完成 LigerSoftmax 算子的昇腾亲和适配，保证算子功能与上游实现一致。

2. 支持 LigerSoftmax 的前向及反向计算，保证前向输出和反向梯度结果正确。

3. 支持 FP16、BF16、FP32 等测试要求的数据类型，确保数值精度满足测试容差。

4. Kernel 实现需符合 Triton-Ascend 编程规范，并针对昇腾 NPU 的计算及访存特性进行适配和优化。

5. 完成适配及优化后，在本地完成全量测试验证与 Benchmark 性能测试，附根因分析、优化手段说明与验证结果，提交 PR 至昇腾算子开源仓。

### 参数说明

| 参数名 | 输入／输出/属性 | 描述 | 数据类型 | 数据格式 | 维度（shape） |
|--------|--------|------|----------|---------|---------------|
| x | 输入 | 输入张量 | FP16/BF16/FP32 | ND | [*] |
| dim | 属性 | 沿指定维度进行 softmax 计算 | INT | - | - |
| out | 输出 | Softmax 计算结果 | 与 x 一致 | ND | [*] |

### 约束说明

1. 输入及输出 Tensor 的 shape、布局、数据类型需与上游 `liger-kernel` LigerSoftmax 接口保持一致。

2. 支持任意维度输入，需具备输入 shape 泛化能力。

3. 前向与反向计算结果均需与上游参考实现（`torch.nn.Softmax`）保持一致。

### 性能优化要求

1. 基于 Benchmark 脚本 `benchmark/scripts/benchmark_softmax.py` 进行性能测试，覆盖 Forward / Backward / Full 三种 `kernel_operation_mode` 配置。

2. 性能目标：所有配置下 Liger Kernel 耗时均不慢于 Baseline（`torch.nn.Softmax` 原生实现）。

3. 优化手段需包含但不限于 Tile/Block 划分策略优化、Kernel Launch 开销优化、非融合路径访存优化及 NPU 指令级优化。

4. 提交 PR 时需附带优化前后逐配置性能对比数据表及 Profiling 分析。

## 测试标准

在 Triton-Ascend 3.2.2 + torch_npu 2.9.0 环境下，运行以下测试：

### 功能测试

```bash
pytest tests/test_softmax.py
```

要求全量用例通过，`0 failed、0 error`，不允许修改测试用例中的断言与容差。

### 性能测试

```bash
python benchmark/scripts/benchmark_softmax.py
```

要求全部配置（forward / backward / full 各 kernel_operation_mode）正常产出数据，无崩溃、无超时。

## 验收标准

1. Pytest 全量用例通过（0 failed、0 error）
2. 数值精度满足测试内 `assert_verbose_allclose` 容差要求
3. Benchmark 全部配置正常产出数据，无崩溃、无超时
4. 性能达标：所有配置 Liger Kernel 耗时均不慢于 Baseline

## PR 申请合入

验收通过后，在 Triton-Ascend-Kernels 开源仓提交 PR 申请，申请将开发完成的算子合入（https://gitcode.com/Ascend/triton-ascend-kernels/tree/experimental）。

## 参考资料

1. [liger-kernel/liger-kernel](https://github.com/linkedin/Liger-Kernel)
2. [Triton-Ascend 文档](https://triton-ascend.readthedocs.io/zh-cn/latest/quick_start.html)
3. [Triton-Ascend 代码样例](https://gitcode.com/Ascend/triton-ascend/tree/main/docs/zh/examples)
4. [Triton-Ascend-Kernels](https://gitcode.com/Ascend/triton-ascend-kernels)
5. 上游 Softmax 测试文件：`tests/test_softmax.py`
6. 上游 Softmax Benchmark：`benchmark/scripts/benchmark_softmax.py`

## 环境获取

依据 [liger-kernel/liger-kernel](https://github.com/linkedin/Liger-Kernel) 仓库说明获取开发环境，配置 Triton-Ascend==3.2.2、torch-npu==2.9.0 及 CANN 9.1.0 环境。

**Comment by lowdy1** [comment]
[softmax.md](https://github.com/user-attachments/files/31590204/softmax.md)

**Comment by HyosungSink** [comment]
@lowdy1 认领这个任务
